packages/joint-core/src/dia/Cell.mjs

```diff
@@ -75,48 +76,58 @@ export const Cell = Model.extend({
+        function removeEmptyAttributes(obj) {
 
-            const defaultAttr = defaultAttrs[selector];
+            for (const key in obj) {
 
-            forIn(attr, function(value, name) {
+                const objValue = obj[key];
+                const isRealObject = isObject(objValue) && !Array.isArray(objValue);
 
-                // attr is mainly flat though it might have one more level (consider the `style` attribute).
-                // Check if the `value` is object and if yes, go one level deep.
-                if (isObject(value) && !Array.isArray(value)) {
+                if (isRealObject) {
+                    removeEmptyAttributes(objValue);
+                }
 
-                    forIn(value, function(value2, name2) {
+                if (isEmpty(objValue) && isRealObject) {
+                    delete obj[key];
+                }
+            }
+        }
 
-                        if (!defaultAttr || !defaultAttr[name] || !isEqual(defaultAttr[name][name2], value2)) {
+        if (ignoreDefaults === false) {
+            // Return all attributes without omitting the defaults
+            const finalAttributes = cloneDeep(this.attributes);
 
-                            finalAttrs[selector] = finalAttrs[selector] || {};
-                            (finalAttrs[selector][name] || (finalAttrs[selector][name] = {}))[name2] = value2;
-                        }
-                    });
+            if (!ignoreEmptyAttributes) return finalAttributes;
 
-                } else if (!defaultAttr || !isEqual(defaultAttr[name], value)) {
-                    // `value` is not an object, default attribute for such a selector does not exist
-                    // or it is different than the attribute value set on the model.
+            removeEmptyAttributes(finalAttributes);
 
-                    finalAttrs[selector] = finalAttrs[selector] || {};
-                    finalAttrs[selector][name] = value;
-                }
+            return finalAttributes;
+        }
+
+        let defaultAttributes = {};
+        let attributes = cloneDeep(this.attributes);
+
+        if (ignoreDefaults === true) {
+            // Compare all attributes with the defaults
+            defaultAttributes = defaults;
+        } else {
+            // Compare only the specified attributes with the defaults, use `attrs` as a default if not specified
+            const differentiateKeys = Array.isArray(ignoreDefaults) ? ignoreDefaults : ['attrs'];
+
+            differentiateKeys.forEach((key) => {
+                defaultAttributes[key] = defaults[key] || {};
             });
-        });
+        }
 
-        const attributes = cloneDeep(omit(this.attributes, 'attrs'));
-        attributes.attrs = finalAttrs;
+        // Omit `type` attribute from the defaults since it should be always present
+        const finalAttributes = objectDifference(attributes, omit(defaultAttributes, 'type'), ignoreEmptyAttributes, 4);
```

Comment:
I would change the signature to `objectDifference(obj, base, { ignoreEmptyAttributes, maxDepth });` and make the function public.

Also does `ignoreEmptyAttributes` ignore the nested attributes too?

Reply:
We should omit `id` attribute too, to be safe.